### ISSUE TITLE
Kestrel dies because of Out of Memory exception

### ISSUE THREAD

**Issue by HassanHashemi** [aspnet/KestrelHttpServer#2840]
Hi, one of our MVC projects keeps growing memory, and it seems that Garbage collector dose not do anything with it.

The process keeps growing until it is killed by the OS.

`dmesg` outputs this log:
`
[961030.365725] Killed process 3673 (dotnet) total-vm:29592364kB, anon-rss:3565416kB, file-rss:256kB, shmem-rss:0kB
`

**What makes me think that it is related to GC not working is that, i have periodically called `GC.Collect() ` and memory immediately drops down.**

Also GC is running under Server mode.
** Environment Info ** 
```
.Net Core 2.1
ASP.NET Core 2.1
Kestrel (directly exposed)
Centos 7

4 Core processor
4GB Memory
```

Also not sure if it any how related, i thought maybe GC dose not recognize memory limits, so here is the output of `ulimit` command:
```
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 15062
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 15062
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited
```

**Comment by benaadams** [comment]
Related to https://github.com/dotnet/coreclr/issues/19060 ? /cc @tmds

**Comment by tmds** [comment]
Are you running this in a container? Enforcing a memory limit?

> dmesg outputs this log: [961030.365725] Killed process 3673 (dotnet) total-vm:29592364kB, anon-rss:3565416kB, file-rss:256kB, shmem-rss:0kB

Is there some more info in dmesg about this?

**Comment by HassanHashemi** [comment]
@tmds No, it's not inside a container. (which is why i think `cgroup` is not the case here). and thus no memory limit other than OS defaults. (as far as i am aware)

> Is there some more info in dmesg about this?
No, `dmesg` contains that one single line of log.

**Comment by tmds** [comment]
Ah, I've seen dmesg OOM logging that was more verbose.

Your system may be out of swap+ram and the kernel is killing a process to free some up.

**Comment by HassanHashemi** [comment]
@tmds What do you think about GC not working?

**Comment by tmds** [comment]
> What do you think about GC not working?

With `anon-rss:3565416kB` and `4GB Memory`, .NET Core still be waiting a bit before GCing.

**Comment by tmds** [comment]
Is this a machine dedicated for running Kestrel? Or does it run other applications/services?

You could install more RAM.
You can also configure the system so it becomes less likely the kernel will kill the dotnet process, but pick another instead. With the default config, probably dotnet looks like an interesting target because it is using a lot of RAM.

**Comment by HassanHashemi** [comment]
It is a dedicated machine for Kestrel.
Adding more RAM is not really an option, because the app dose not really need that much. `dotnet` should not be using that amount of memory anyway.

**Comment by tmds** [comment]
I'm surprised no other things are running on the machine.
It's unfortunate dmesg has so little information why the process is killed.

> dotnet should not be using that amount of memory anyway.

the amount of physical ram is used to determine how much memory dotnet can use.
If you want to lower it, you can run it in a cgroup.

**Comment by Drawaes** [comment]
If you run it as workstation gc it will be more agressive. How many cores is the machine because server gc has a heap per core so can easily swallow a lot of ram as it's baseline before gcing. Workstation will be more agressive (at the cost of more gc activity)

**Comment by HassanHashemi** [comment]
> the amount of physical ram is used to determine how much memory dotnet can use.

as an example, with 4GB memory dotnet should not keep allocating memory, because the limit is already set by amount of physical memory, so why do we need to enforce the limit via cgroup?

**Comment by HassanHashemi** [comment]
@Drawaes Yeah, the machine is 4 has 4 Cores and is running under ServerGC. (i believe having ServerGC enabled is a reasonable thing to have)

**Comment by benaadams** [comment]
Might want to raise an issue in https://github.com/dotnet/coreclr

**Comment by Drawaes** [comment]
That is less than 1gb per core (minus os overhead) which is pretty low for the server gc.

The default GC segment sizes aren't going to be good for you.

https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals#ephemeral-generations-and-segments

**Comment by HassanHashemi** [comment]
@Drawaes ah i see, let me add 4 GB to the machine and observe the result.

**Comment by tmds** [comment]
> as an example, with 4GB memory dotnet should not keep allocating memory, because the limit is already set by amount of physical memory, so why do we need to enforce the limit via cgroup?

By default it will take into account the physical memory limit. If you _want_ to force dotnet to a lower value, you _can_ use a cgroup.

> I have deployed the app again, and this time i will also monitor swap info.

To figure out why the system does an OOM kill. These are interesting: free memory, free swap and the dotnet proc status file.

> let me add 4 GB to the machine and observe the result.

If this reproduces well, let's try and figure out the root cause before adding more RAM.

**Comment by tmds** [comment]
@HassanHashemi I noticed you included another line in the `dmesg` output. The full output I was expecting was something like https://stackoverflow.com/questions/9199731/understanding-the-linux-oom-killers-logs. This includes a lot of information on where the memory is spent.

**Comment by HassanHashemi** [comment]
@tmds yeah, sorry. that was missed in copy+paste phase.

**Comment by tmds** [comment]
I've checked: .NET Core is properly detecting the amount of physical RAM as a limit on my system. So probably your app has a memory leak.

**Comment by HassanHashemi** [comment]
@tmds If that's the case, why would explicitly calling `GC.Collect()` frees up the memory?

**Comment by Drawaes** [comment]
A simple test is set it to workstation gc. If you don't get the oom crash then it's not a leak because it shouldn't be able to clean up a leak.

**Comment by tmds** [comment]
> @tmds If that's the case, why would explicitly calling GC.Collect() frees up the memory?

Calling GC.Collect may be postponing the OOM in case of a memory leak.
How much memory does it free up?

**Comment by tmds** [comment]
If you want to know what .NET Core is using as a memory limit you can do this:
```sh
$ wget https://raw.githubusercontent.com/dotnet/coreclr/master/src/gc/unix/cgroup.cpp
$ cat >main.cpp <<EOF
#include "cgroup.cpp"
int main()
{
  printf("%zu\n", GetRestrictedPhysicalMemoryLimit());
  return 0;
}
EOF
$ g++ main.cpp
$ ./a.out
16814272512
```
For me this is 16GB.

This value is used by the GC. The GC is not aware where the limit comes from (cgroup, amount of RAM, ...).

**Comment by HassanHashemi** [comment]
Thanks @tmds i have followed what you pointed out and found that GC is running correctly.

However after investigating with MemProfiler, there `ModelExplorer` is keeping some objects, here is the graph representing the GC path.

![image](https://user-images.githubusercontent.com/6805984/44625465-a0b38500-a91f-11e8-9864-ad2b065e2bad.png)

Any ideas is highly appreciated.

**Comment by benaadams** [comment]
That's the Redis ResultBox cache rooting the objects; though will keep at max 64 items. Have added PR for it https://github.com/StackExchange/StackExchange.Redis/pull/927

**Comment by davidfowl** [comment]
This diagram helps, thanks @HassanHashemi ! It looks like we have a big problem with the interaction between the IHttpContextAccessor and the way we use ChangeTokens to cache view information. 

cc @pranavkm

**Comment by davidfowl** [comment]
@HassanHashemi is this during development or are you seeing these types of issues in production?

**Comment by HassanHashemi** [comment]
@davidfowl this happened on the production server (CentOS).

**Comment by davidfowl** [comment]
Interesting, do you have cshtml files on your production server?

**Comment by HassanHashemi** [comment]
@davidfowl No, they are built into library. (Project.Views.dll)

**Comment by HassanHashemi** [comment]
Which confused me also, i was expecting `DefaultRazorEngineFileProviderAccessor` to have this bahavior on cshtml files. it kind of dose not make sense to have FileSystemWatcher for InMemory Views.
This led me to check if it's loading *.cshtml files. (a few more info [here](https://github.com/benaadams/Ben.BlockingDetector/issues/10))

**Comment by benaadams** [comment]
Some of the paths should be fixed (right hand side, Sql/DbConnectionPool etc) in the .NET Core 2.2 preview https://blogs.msdn.microsoft.com/webdev/2018/08/22/asp-net-core-2-2-0-preview1-now-available/

**Comment by HassanHashemi** [comment]
@benaadams Why? any related details?

**Comment by benaadams** [comment]
>@benaadams Why? any related details?

Removed capturing of asynclocals into lazy initalized static timers; such as the ConnectionPools
https://github.com/dotnet/corefx/pulls?utf8=%E2%9C%93&q=is%3Apr+author%3Abenaadams+is%3Aclosed+asynclocals

**Comment by benaadams** [comment]
Hmm.. that might have been in 2.1.x?

**Comment by HassanHashemi** [comment]
It's already 4.5.1 (together with EF Core 2.1.1)

**Comment by davidfowl** [comment]
@HassanHashemi are you using Razor Pages or MVC views (It doesn't matter I'm just curious)?

**Comment by HassanHashemi** [comment]
MVC views only.

**Comment by davidfowl** [comment]
I'd be surprised if this alone was responsible for the OOM. Even if you have issues with rooting the  HttpContext, you'd have one per file or directory and that's a finite set. What else does the profile show?

**Comment by HassanHashemi** [comment]
@davidfowl I only followed the roots that were part of my code and were expected to be reclaimed by GC.
I honestly have no odea what types in Aspnet core namespace should be followed too. 

> I'd be surprised if this alone was responsible for the OOM. Even if you have issues with rooting the HttpContext, you'd have one per file or directory and that's a finite set. What else does the profile show?

The above objects at top of this tree grow in count as number of request go up. This app handles 100 request per second and fill 4GB in 3 days.

**Comment by HassanHashemi** [comment]
@Tratcher Is this going to be fixed any time soon?

**Comment by Tratcher** [comment]
This wasn't a single issue, it was a whole category of issues. Several linked work items were created and some of those have already been fixed in the 2.2.0 previews. https://github.com/aspnet/Mvc/issues/8362 looks most directly related to your last post.

**Comment by davidfowl** [comment]
We addressed it in 2.2, we can consider patching this in 2.1 but it might be too big of a change to do that. I'll discuss it with some people.

cc @muratg @Eilon @DamianEdwards

**Comment by guohuang** [comment]
> Are you running this in a container? Enforcing a memory limit?
> 
> > dmesg outputs this log: [961030.365725] Killed process 3673 (dotnet) total-vm:29592364kB, anon-rss:3565416kB, file-rss:256kB, shmem-rss:0kB
> 
> Is there some more info in dmesg about this?

@tmds we are having a similar issue in a container, the Kestrel is not responding to a ping service. (service timeout) which kills the container.  we have set soft limit 512mb and hard limit 1gb. 

at this point, we are not 100% sure if it is memory issue, the memory usage is ~70%, but we notice when the server stopped responding, our thread count is 150+. (when the app is idle, it is ~40).

@tmds do you have any suggestions on diagnosing the issue?

**Comment by halter73** [comment]
150+ threads is a lot for an app you're trying to limit to 512mb/1gb of memory. Most linux distros reserver 8mb of virtual memory per thread/stack by default. It seems like you might be experiencing a problem related to threadpool starvation. If you're not sure what's consuming all of these threads, you can investigate by taking a memory dump (the createdump utility next to libcoreclr.so is useful for this) and [examining it with lldb/sos](https://github.com/dotnet/coreclr/blob/85ed652fef5f1dec7c532bd4963dd3cde0199211/Documentation/building/debugging-instructions.md#debugging-core-dumps-with-lldb).

**Comment by benaadams** [comment]
@HassanHashemi how does it behave in the newly released 2.2?

**Comment by HassanHashemi** [comment]
@benaadams I have tested this locally under Profiler and it was completely fixed.
Will try it in production later this week.
